### PR TITLE
13065 Mobile Auth

### PR DESCRIPTION
**Pull request by kreek**
## Description of change
To split these changes up there's a PR with the core vets-api `ApplicationController` changes [here](https://github.com/department-of-veterans-affairs/vets-api/pull/4914). Please look at that PR first.

Implements the authentication flow to the mobile engine. Adds an authenticate method to the mobile `ApplicationController` which uses the `IAMSSOeOAuth::SessionManager` to find and create sessions with it's own versions of the user, user_identity, and session classes (which all inherit from the core vets-api versions of those classes).

Also adds the initial version of the `/mobile/v0/user` endpoint. This will return more user traits once the mobile screens are finalized.

## Original issue(s)
department-of-veterans-affairs/va.gov-team#13065

## Things to know about this PR
* Adds redis.yml entries for IAMUser, IAMUserIdentity, IAMSession classes
* The mobile auth design doc is [here](https://github.com/department-of-veterans-affairs/va.gov-team/blob/mobile-ssoe-design-doc/platform/engineering/design-docs/2020-08-12-mobile-ssoe-auth-design-doc.md)

### CODE REVIEW

File: app/models/iam_user_identity.rb [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+  #
+  # @param iam_profile [Hash] the profile of the user, as they are known to the IAM SSOe service.
+  # @return [IAMUserIdentity] an instance of this class
+  #
+  def self.build_from_iam_profile(iam_profile)
+    loa_level = iam_profile[:fediamassur_level].to_i
+
+    identity = new(
+      email: iam_profile[:email],
+      expiration_timestamp: iam_profile[:exp],
```
Comment (jjulian) [inline app/models/iam_user_identity.rb:26]: Can we confirm that the format of `exp` is compatible with the call to `redis_namespace.expireat`?

File: app/models/iam_user_identity.rb [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+      first_name: iam_profile[:given_name],
+      iam_icn: iam_profile[:fediam_mviicn],
+      iam_edipi: iam_profile[:fediam_do_dedipn_id],
+      iam_sec_id: iam_profile[:fediamsecid],
+      last_name: iam_profile[:family_name],
+      loa: { current: loa_level, highest: loa_level },
```
Comment (jjulian) [inline app/models/iam_user_identity.rb:32]: I think this is ok, but want to confirm. loa_level will always be "3" from IAM SSOe?

File: config/redis.yml [hunk 3]
```diff
@@ -102,6 +102,15 @@ development: &defaults
...
     each_ttl: 3600 # 1 hour
+  iam_session:
+    namespace: iam-session
+    each_ttl: 1800
+  iam_user:
+    namespace: iam-user
+    each_ttl: 1800
+  iam_user_identity:
+    namespace: iam-user-identity
+    each_ttl: 1800
```
Comment (jjulian) [inline config/redis.yml:113]: Is this 30 minutes to match the access_token ttl?

File: modules/mobile/spec/spec_helper.rb [hunk 4]
```diff
@@ -1,15 +0,0 @@
...
-
-require 'rspec/rails'
-
-ENGINE_RAILS_ROOT = File.join(File.dirname(__FILE__), '../')
-
-# Requires supporting ruby files with custom matchers and macros, etc,
-# in spec/support/ and its subdirectories.
-Dir[File.join(ENGINE_RAILS_ROOT, 'spec/support/**/*.rb')].sort.each { |f| require f }
-
-RSpec.configure { |config| config.use_transactional_fixtures = true }
```
Comment (jjulian) [inline modules/mobile/spec/spec_helper.rb:15]: `spec_helper.rb` is not needed?

Comment (kreek) [reply modules/mobile/spec/spec_helper.rb:15]: Realized we want to use the core vets-api spec_helper so this one's not needed.

File: config/redis.yml (hunk 3 above)
Comment (kreek) [reply config/redis.yml:113]: Yes, but this is a default. The TTL will be set from the introspect `exp` value.

File: app/models/iam_user_identity.rb (hunk 2 above)
Comment (kreek) [reply app/models/iam_user_identity.rb:33]: 👍 I'll double check that with Perry.

File: app/models/iam_user_identity.rb (hunk 1 above)
Comment (kreek) [reply app/models/iam_user_identity.rb:27]: Redis namespace gem aliases all the Redis methods and expireat takes a unix timestamp 
https://github.com/redis/redis-rb/blob/master/lib/redis.rb#L430-L439

**Review by jjulian — APPROVED** [review]
Looks great!

**Comment by kfrz** [comment]
Working to review this, though there are a few build failures it seems and merge resolutions required. 

Exciting stuff!

**Comment by kreek** [comment]
@kfrz this is passing CI now, I did fix something but there was also a VBA Documents spec error which seems dependent on the run order.

**Comment by erluti** [comment]
> a VBA Documents spec error which seems dependent on the run order.

This test flaps sometimes and we're trying to figure out why.  Do you have some more evidence for this?

**Comment by kreek** [comment]
@erluti the failing VBA Document run

http://jenkins.vfs.va.gov/blue/organizations/jenkins/testing%2Fvets-api/detail/PR-4898/18/pipeline

File: modules/mobile/app/serializers/mobile/v0/user_serializer.rb [hunk 5]
```diff
@@ -0,0 +1,17 @@
+# frozen_string_literal: true
+
+require 'fast_jsonapi'
+
+module Mobile
+  module V0
+    class UserSerializer
+      include FastJsonapi::ObjectSerializer
+
+      # TODO: return full set of user data, including latest appeals, claims, and appointments
```
Comment (kfrz) [inline modules/mobile/app/serializers/mobile/v0/user_serializer.rb:10]: Is this `TODO` before the PR is merged, or is there a follow-up ticket for traceability purposes?

File: modules/mobile/spec/controllers/application_controller_spec.rb [hunk 6]
```diff
@@ -0,0 +1,133 @@
+# frozen_string_literal: true
+
+require 'rails_helper'
+require_relative '../rails_helper'
+
```
Comment (kfrz) [inline modules/mobile/spec/controllers/application_controller_spec.rb:5]: Why is this necessary?

**Review by kfrz — APPROVED** [review]
This looks good to me, just a couple of nitpick/questions,  `TODO`s shouldn't be left in code committed, IMO, without an issue linked in the comment as well

Otherwise, :clap:

File: modules/mobile/app/serializers/mobile/v0/user_serializer.rb (hunk 5 above)
Comment (kreek) [reply modules/mobile/app/serializers/mobile/v0/user_serializer.rb:10]: this is for follow up, wanted a slimed down version to test auth before the full user endpoint goes in another PR

File: modules/mobile/spec/controllers/application_controller_spec.rb (hunk 6 above)
Comment (kreek) [reply modules/mobile/spec/controllers/application_controller_spec.rb:5]: I should rename the second one, the first one is the core 'rails_helper', the second is the module's 'rails_helper' which has Rspec configurations specific to this module, e.g. flipping the feature on, creating a signed in IAM user.

File: modules/mobile/app/serializers/mobile/v0/user_serializer.rb (hunk 5 above)
Comment (kreek) [reply modules/mobile/app/serializers/mobile/v0/user_serializer.rb]: added and issue # to the todo
